### ISSUE TITLE
The result of comparison operations may have requires_grad=True

### ISSUE THREAD

**Issue by colesbury** [pytorch/pytorch#3348]
For example:

```python
import torch
from torch.autograd import Variable
v = Variable(torch.randn(5,5), requires_grad=True)
x = v < 0
print(x.requires_grad)  # True
```

The problem is that many functions now check that their Byte/LongTensor inputs  **don't** require grad.

**Comment by colesbury** [comment]
https://discuss.pytorch.org/t/get-error-message-maskedfill-cant-differentiate-the-mask/9129/4

**Comment by colesbury** [comment]
Removed 0.3 blocker tag. We should fix this in `master` but I think the referenced behavior is true in 0.2.

**Comment by gchanan** [comment]
This was fixed.
